### ISSUE TITLE
Add `ur_device_handle_t` parameter to `urEnqueueUSMMemAdvise`

### ISSUE THREAD

**Issue by igchor** [oneapi-src/unified-runtime#336]
`ur_mem_advice_t` specifies: `UR_MEM_ADVICE_SET_PREFERRED_LOCATION` and `UR_MEM_ADVICE_CLEAR_PREFERRED_LOCATION` but `urEnqueueUSMMemAdvise` does not accept any device handle which makes it impossible to specify the location.

**Comment by igchor** [comment]
Extending `urUsm*Alloc` function might not be necessary anymore. Probably we can just use mechanism introduced in https://github.com/oneapi-src/unified-runtime/pull/359 (separate descriptors for host/device/shared) to set PREFFERED_LOCATION.

**Comment by igchor** [comment]
I think this issue might be related to https://github.com/oneapi-src/unified-runtime/issues/423 in regard to how we want to expose the advice hints (just a single uin32_t or separate `desc` types for host/shared/device as for flags).

@jandres742 I was also wondering about https://spec.oneapi.io/level-zero/latest/core/api.html?highlight=advice#zecommandlistappendmemadvise Can `hDevice` that is passed to this function be different than `hDevice` passed to `zeMemAllocShared`? Would it make sense to associated device `A` with shared allocation but set `PREFERRED_LOCATION` to a device `B`?

**Comment by jandres742** [comment]
@igchor : in L0, a shared-allocation may migrate to other devices different to the one specified on allocation. So in that case, memAdvise for other devices would make sense.

**Comment by kbenzie** [comment]
> I think this issue might be related to #423 in regard to how we want to expose the advice hints (just a single uin32_t or separate `desc` types for host/shared/device as for flags).

Hopefully the change I made makes sense 😬

**Comment by igchor** [comment]
> > I think this issue might be related to #423 in regard to how we want to expose the advice hints (just a single uin32_t or separate `desc` types for host/shared/device as for flags).
> 
> Hopefully the change I made makes sense 😬

Yes, for hints we do not have any extra state/context so no need to expose extra descriptors for host/device.

I was just trying to understand what we need for PREFERRED_LOCATION but it looks like we only need to add `hDevice` to urEnqueueUSMMemAdvise and nothing needs to be changed for the allocation functions (we can still pass PREFERRED_LOCATION through a descriptor and make the specified device preferred.

**Comment by igchor** [comment]
I think I agree that having the `_DEVICE` suffix might be better - especially since we can pass hints to both `urEnqueueUSMAdvise` and allocation functions (through a descriptor).
